### DISCUSSION TITLE
Generating game engine boilerplate

### DISCUSSION THREAD

**Discussion by patrickdown** [hsutter/cppfront#879] [Q&A] [answered]
I've been following this project with interest. One thing I do is work  with C++ and game engines. One thing that a they have in common is boilerplate that a class needs to implement to work in the engine. This is often accomplished with macros and sometimes with custom code generation passes.

In cppfront I could see parts of this being accomplish with custom metafuctions.

```
MyNode: @godot_class type = {
	this: Node;
	// ...
} 
```

But there is another part that I'm not sure how the current cppfront would help with the boilerplate.  This involves exporting properties/methods to be visible in the editor or scripting language.

Example from Godot that uses template magic and macros.
```
void VoxelMesh::_bind_methods() {
    ClassDB::bind_method(D_METHOD("set_volume_size", "size"), &VoxelMesh::set_volume_size);
    ClassDB::bind_method(D_METHOD("get_volume_size"), &VoxelMesh::get_volume_size);
    ADD_PROPERTY(PropertyInfo(Variant::VECTOR3I, "volume_size", godot::PROPERTY_HINT_NONE), "set_volume_size", "get_volume_size");
    // ...
}
```

It would be nice if the `@godot_class` metafunction could build the `_bind_methods` for you but how does it identify the methods? 

Custom attributes that could be queried by the metafunction?
```
[godot_export] set_volume_size: (size : Vector3i) = {}
```

A tagging though the method name where the name is rewritten? :scream:
```
gmethod_set_volume_size: (size : Vector3i) = {}
```

**Comment by MaxSagebaum** [comment]
From the error messages in the code Herb wants to make metafunctions available on nearly everything. https://github.com/hsutter/cppfront/blob/2d9382db3681e7df4ef5f602aa20fb7e19aaf32d/source/parse.h#L8050

So it would be possible to write:
```
MyNode: @godot_class type = {
	this: Node;
       
        volume_size: @godot_property VECTOR3I = {};
}
```

**Reply by JohelEGP** [reply, marked as answer]
Even with only a type metafunction, you can use an alias for `std::type_identity_t` named `godot_property` that `@godot_class` can observe.

**Reply by patrickdown** [reply]
This seems like a good short term solution thanks,
